ActiveJob: Improve compatibility with Zeitwerk

pboling opened the pull request:
Following are removed because `active_support` autoloads them:

    require "active_support/tagged_logging"
    require "active_support/logger"

In accordance with Rails Guides' documentation on how to load active_support, and in compliance with patterns (the good ones, not the bad ones) everywhere else in Rails code.

- See: https://guides.rubyonrails.org/active_support_core_extensions.html

### Motivation / Background

According to documentation [here on Rails Guides](https://guides.rubyonrails.org/active_support_core_extensions.html)
libraries that are not `active_support` that depend on parts of `active_support` should first do:

```
require "active_support"
```

and then whatever specific part they need.

Or, if they need the whole thing they should instead do:
```
require "active_support/all"
```

They should never drop in like a bird of prey and pluck out some internal part of `active_support` without first doing:

```
require "active_support"
```

Unfortunately this is what [active_job does](https://github.com/rails/rails/blob/main/activejob/lib/active_job/logging.rb#L3-L4), and it breaks some libraries I am working on.

Currently the top of `active_job/logging.rb` has:

```
require "active_support/tagged_logging"
require "active_support/logger"
```

It should be changed to:

```
require "active_support"
```

Since TaggedLogging and Logger will be loaded by `autoload`.

This affects every version of Rails that has active_job, tested with 7.2 and 8.0.

Tellingly, it is the only violation of the pattern set forth in the Rails Guides in the entire Rails code base, which is a strong indicator that it should be fixed.

### Detail

This Pull Request changes `activejob/lib/active_job/logging.rb`

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

pboling commented:
The buildkite/rails failure does not appear to be related in any way to this change.  Perhaps the test ENV is flaky?

skipkayhil commented on activejob/lib/active_job/logging.rb:
```diff
@@ -1,7 +1,6 @@
 
-require "active_support/tagged_logging"
-require "active_support/logger"
+require "active_support"
```
`require "active_support"` already happens in the Active Job entrypoint, so it should not be added here.

https://github.com/rails/rails/blob/800976975253be2912d09a80757ee70a2bb1e984/activejob/lib/active_job.rb#L26

However, you are right that these requires can be removed because they reference an autoloaded constant (`TaggedLogging`) and an already required file (`active_support.rb` requires `active_support/logger`)

pboling replied:
I agree.  It seems like there are historical patterns around the codebase of Rails which end up getting copied and pasted into other code, even though the paradigm has been dead for a long time.

I've pushed the branch again with the removal of that line as well.

pboling committed:
Improve compatibility with autoload

Following are removed because `active_support` autoloads them:

    require "active_support/tagged_logging"
    require "active_support/logger"

In accordance with Rails Guides' documentation on how to load active_support, and in compliance with patterns everywhere else in Rails code.

- Fixes rails#54639
- See: https://guides.rubyonrails.org/active_support_core_extensions.html

pboling commented:
These are far from the only lines that exhibit requires that would throw a wrench into Zeitwerk, but since they're pre-existing I think removing them one file at a time is fine.

pboling commented:
The issue is that by hardcoding those requires it preempts Zeitwerk, and thus prevents other things that are very nice to have.

Rails currently only pretends to use Zeitwerk.

rafaelfranca commented:
Can you give me an example why it preempts Zeitwerk? I can't see how.

pboling commented:
The classes can't be autoloaded if they have been loaded by a require.

The beauty of autoload is that I can define the class in another gem, keeping the same API, and essentially swap out internals as needed.

These hard coded requires prevent me from doing that.

pboling commented:
I'm writing an extraction of the Rails logging toolchain where I can test and fix issues in the standard Rails logging toolchain.  Tagged Logging with Broadcasting has been fundamentally broken for a long time:
  
- https://github.com/rails/rails/issues/49757
- https://github.com/rails/rails/issues/49745
- https://github.com/rails/rails/issues/46084
- https://github.com/rails/rails/issues/44668

So I created these:

- Enhanced [activesupport-logger](https://github.com/pboling/activesupport-logger) which was ripped from Rails v8.0
- Enhanced [activesupport-broadcast_logger](https://github.com/pboling/activesupport-broadcast_logger) which was ripped from Rails v8.0, and https://github.com/rails/rails/pull/53093
- Enhanced [activesupport-tagged_logging](https://github.com/pboling/activesupport-tagged_logging) which was ripped from Rails v8.0, and https://github.com/rails/rails/pull/53105

As test beds for proposed fixes.  I also added back support for all versions of Rails back to 5.2, so they should be able to unify logging config (and crucially output) across a deployment of many discrete Rails apps running myriad versions of Rails.

It's working, except we still haven't migrated from Sidekiq job classes to ActiveJob, and those two lines are preventing us from switching.

PS. If you're thinking "why would anyone..." I can relate.  In this particular industry (FDA regulated-and-audited medical devices) being on a "fork of Rails" would require much more validation than being on a standard Rails with a "gem fork of the Rails' logger".

rafaelfranca commented:
Rails has different requirements of Rails applications. For example, you don't want to eager load the postgresql adapter in an application using mysql. Using zeitlwerk, would require us to white custom logic for this kind of requirements, so using zeitwerk in some of the frameworks doesn't bring much advantages.



> The classes can't be autoloaded if they have been loaded by a require.
> 
> The beauty of autoload is that I can define the class in another gem, keeping the same API, and essentially swap out internals as needed.
> 
> These hard coded requires prevent me from doing that.

This has nothing to do with Zeitwerk. What you are describing is that you want a way to monkey patch Rails by using autoload. All I can say is that isn't supported and never will.

rafaelfranca commented:
That being said. I makes me sad that you need to create all those gems. I'll make sure those issues are fixed, and I hope that helps you as well.

pboling commented:
I'd love to see movement on the logging bugs!

Regarding autoload, the dream is the one inherited from the merb 3 merge.  We could have a truly modular Rails if more parts of Rails relied exclusively on autoload.  I think that would be a pretty good thing, and enable some really interesting abilities.
